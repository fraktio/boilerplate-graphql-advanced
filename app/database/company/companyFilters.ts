import { Maybe } from "~/@types/global";
import { FilterOperator } from "~/database/filters/operators";
import { StringFilter } from "~/database/filters/stringFilters";

export type CompanyFilter = {
  filterOperations?: Maybe<Array<CompanyFilterOperation>>;
  nameFilter?: Maybe<StringFilter>;
};

export type CompanyFilterOperation = {
  operator: FilterOperator;
  filters?: Maybe<Array<CompanyFilter>>;
};
