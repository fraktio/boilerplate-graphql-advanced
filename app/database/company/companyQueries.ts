import { Knex } from "knex";
import { DateTime } from "luxon";

import { DBSession } from "~/database/connection";
import { applyStringFilters, buildFilterQuery } from "~/database/filters";
import { createUUID, ID, Table, tableColumn } from "~/database/tables";
import {
  CompanyFilter,
  CompanyFilterOperation,
  FilterOperator,
  Maybe,
} from "~/generation/generated";
import { UUID } from "~/generation/mappers";

export interface CompanyID extends ID {
  __CompanyID: never;
}

export type CompanyTableRow = Readonly<{
  id: CompanyID;
  uuid: UUID;
  name: string;
  createdAt: Date;
  updatedAt: Date | null;
}>;

export type CompanyTable = {
  id: CompanyID;
  UUID: UUID;
  name: string;
  timestamp: {
    createdAt: DateTime;
    updatedAt: DateTime | null;
  };
};

export type CreateCompanyParams = {
  name: string;
};

export const formatCompanyRow = (row: CompanyTableRow): CompanyTable => ({
  id: row.id,
  UUID: row.uuid,
  name: row.name,
  timestamp: {
    createdAt: DateTime.fromJSDate(row.createdAt),
    updatedAt: row.updatedAt ? DateTime.fromJSDate(row.updatedAt) : null,
  },
});

export const companyQueries = {
  async get(params: {
    knex: DBSession;
    companyId: CompanyID;
  }): Promise<Maybe<CompanyTable>> {
    const company = await params
      .knex<CompanyTableRow>(Table.COMPANY)
      .where({ id: params.companyId })
      .first();

    return company ? formatCompanyRow(company) : null;
  },

  async tryGet(params: {
    knex: DBSession;
    companyId: CompanyID;
  }): Promise<CompanyTable> {
    const company = await companyQueries.get({
      knex: params.knex,
      companyId: params.companyId,
    });

    if (!company) {
      throw new Error("Invalid companyId");
    }

    return company;
  },

  async getByUUID(params: {
    knex: DBSession;
    companyUUID: UUID;
  }): Promise<Maybe<CompanyTable>> {
    const company = await params
      .knex<CompanyTableRow>(Table.COMPANY)
      .where({ uuid: params.companyUUID })
      .first();

    return company ? formatCompanyRow(company) : null;
  },

  async getByIds(params: {
    knex: DBSession;
    companyIds: CompanyID[];
  }): Promise<CompanyTable[]> {
    const persons = await params
      .knex<CompanyTableRow>(Table.COMPANY)
      .whereIn("id", params.companyIds);

    return persons.map(formatCompanyRow);
  },

  async getAll(params: {
    knex: DBSession;
    filters?: CompanyFilterOperation;
  }): Promise<CompanyTable[]> {
    const companies = await params
      .knex<CompanyTableRow>(Table.COMPANY)
      .andWhere((qb) => addCompanyFilters(qb, params.filters));

    return companies.map(formatCompanyRow);
  },

  async create(params: {
    knex: DBSession;
    company: CreateCompanyParams;
  }): Promise<CompanyTable> {
    const companies = await params
      .knex<CompanyTableRow>(Table.COMPANY)
      .insert({
        uuid: createUUID(),
        name: params.company.name,
      })
      .returning("*");

    return formatCompanyRow(companies[0]);
  },

  async updateByUUID(params: {
    knex: DBSession;
    companyUUID: UUID;
    company: { name: string };
  }): Promise<Maybe<CompanyTable>> {
    const companies = await params
      .knex<CompanyTableRow>(Table.COMPANY)
      .update({ name: params.company.name })
      .where({ uuid: params.companyUUID })
      .returning("*");

    if (companies.length === 0) {
      return null;
    }

    return formatCompanyRow(companies[0]);
  },

  async getCompaniesByIds(params: {
    knex: DBSession;
    companyIds: readonly CompanyID[];
  }): Promise<CompanyTable[]> {
    const companyRows = await params
      .knex<CompanyTableRow>(Table.COMPANY)
      .whereIn("id", params.companyIds);

    return companyRows.map(formatCompanyRow);
  },
};

export function addCompanyFilters(
  queryBuilder: Knex.QueryBuilder,
  filterOperation?: CompanyFilterOperation,
): Knex.QueryBuilder {
  return buildFilterQuery(queryBuilder, applyCompanyFilters, filterOperation);
}

function applyCompanyFilters(input: {
  queryBuilder: Knex.QueryBuilder;
  filterOperator: FilterOperator;
  filter: CompanyFilter;
}): Knex.QueryBuilder {
  const { queryBuilder, filterOperator, filter } = input;

  if (filter.nameFilter) {
    return applyStringFilters({
      queryBuilder,
      filterOperator,
      field: tableColumn(Table.COMPANY, "name"),
      stringFilter: filter.nameFilter,
    });
  }

  return queryBuilder;
}
