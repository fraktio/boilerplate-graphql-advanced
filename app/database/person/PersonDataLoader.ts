import DataLoader from "dataloader";

import {
  AbstractDataLoaderBase,
  DataLoaderParams,
} from "~/database/AbstractDataLoader";
import {
  personQueries,
  PersonID,
  PersonTable,
} from "~/database/person/personQueries";
import { Maybe } from "~/generation/generated";

export type PersonLoader = DataLoader<PersonID, Maybe<PersonTable>>;

export class PersonDataLoader extends AbstractDataLoaderBase<PersonLoader> {
  protected createLoader(params: DataLoaderParams): PersonLoader {
    const personLoader: PersonLoader = new DataLoader(async (ids) => {
      const employees = await personQueries.getPersonsByIds({
        knex: params.knex,
        personIds: ids,
      });

      return ids.map(
        (id) => employees.find((employee) => employee.id === id) || null,
      );
    });

    return personLoader;
  }
}
