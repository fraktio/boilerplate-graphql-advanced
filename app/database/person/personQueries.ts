import { Knex } from "knex";
import { PhoneNumber, parsePhoneNumber } from "libphonenumber-js";
import { DateTime } from "luxon";

import { DBSession } from "~/database/connection";
import {
  buildFilterQuery,
  applyDateFilters,
  applyStringFilters,
} from "~/database/filters";
import { createUUID, ID, Table, tableColumn } from "~/database/tables";
import {
  Maybe,
  PersonFilterOperation,
  PersonFilter,
  FilterOperator,
  PersonSort,
  SortOrder,
} from "~/generation/generated";
import { UUID } from "~/generation/mappers";
import {
  CountryCode,
  EmailAddress,
  FinnishPersonalIdentityCode,
} from "~/generation/scalars";
import { asCountryCode } from "~/validation/converters";

export interface PersonID extends ID {
  __PersonID: never;
}

export type PersonTableRow = Readonly<{
  id: PersonID;
  uuid: UUID;
  firstName: string;
  lastName: string;
  phone: string;
  email: EmailAddress;
  birthday: Date;
  createdAt: Date;
  updatedAt: Date | null;
  nationality: CountryCode;
  personalIdentityCode: FinnishPersonalIdentityCode;
}>;

export type PersonTable = {
  id: PersonID;
  UUID: UUID;
  firstName: string;
  lastName: string;
  phone: PhoneNumber;
  email: EmailAddress;
  birthday: DateTime;
  timestamp: {
    createdAt: DateTime;
    updatedAt: DateTime | null;
  };
  nationality: CountryCode;
  personalIdentityCode: FinnishPersonalIdentityCode;
};

export const formatPersonRow = (row: PersonTableRow): PersonTable => ({
  id: row.id,
  UUID: row.uuid,
  firstName: row.firstName,
  lastName: row.lastName,
  phone: parsePhoneNumber(row.phone),
  email: row.email,
  birthday: DateTime.fromJSDate(row.birthday),
  nationality: asCountryCode(row.nationality),
  personalIdentityCode: row.personalIdentityCode,
  timestamp: {
    createdAt: DateTime.fromJSDate(row.createdAt),
    updatedAt: row.updatedAt ? DateTime.fromJSDate(row.updatedAt) : null,
  },
});

export type CreatePersonOptions = {
  firstName: string;
  lastName: string;
  phone: PhoneNumber | null;
  email: EmailAddress;
  birthday: DateTime;
  nationality: CountryCode;
  personalIdentityCode: FinnishPersonalIdentityCode;
};

export type UpdatePersonOptions = CreatePersonOptions;

export const personQueries = {
  async get(params: {
    knex: DBSession;
    id: PersonID;
  }): Promise<Maybe<PersonTable>> {
    const person = await params
      .knex<PersonTableRow>(Table.PERSONS)
      .where({ id: params.id })
      .first();

    return person ? formatPersonRow(person) : null;
  },

  async getByUUID(params: {
    knex: DBSession;
    personUUID: UUID;
  }): Promise<Maybe<PersonTable>> {
    const person = await params
      .knex<PersonTableRow>(Table.PERSONS)
      .where({ uuid: params.personUUID })
      .first();

    return person ? formatPersonRow(person) : null;
  },

  async getAll(params: {
    knex: DBSession;
    filters?: PersonFilterOperation;
    sort?: PersonSort[];
  }): Promise<PersonTable[]> {
    const persons = await params
      .knex<PersonTableRow>(Table.PERSONS)
      .andWhere((qb) => addPersonFilters(qb, params.filters))
      .orderBy(applyPersonSort(params.sort));

    return persons.map(formatPersonRow);
  },

  async create(params: {
    knex: DBSession;
    person: CreatePersonOptions;
  }): Promise<PersonTable> {
    const phone = params.person.phone?.formatInternational();
    const birthday = params.person.birthday.toJSDate();

    const persons = await params
      .knex<PersonTableRow>(Table.PERSONS)
      .insert({
        uuid: createUUID(),
        firstName: params.person.firstName,
        lastName: params.person.lastName,
        phone,
        email: params.person.email,
        birthday,
      })
      .returning("*");

    return formatPersonRow(persons[0]);
  },

  async updateByUUID(params: {
    knex: DBSession;
    personUUID: UUID;
    person: UpdatePersonOptions;
  }): Promise<Maybe<PersonTable>> {
    const phone = params.person.phone?.formatInternational();
    const birthday = params.person.birthday.toJSDate();

    const persons = await params
      .knex(Table.PERSONS)
      .update({
        firstName: params.person.firstName,
        lastName: params.person.lastName,
        phone,
        email: params.person.email,
        birthday,
      })
      .where({ uuid: params.personUUID })
      .returning("*");

    if (persons.length === 0) {
      return null;
    }

    return formatPersonRow(persons[0]);
  },

  async getPersonsByIds(params: {
    knex: DBSession;
    personIds: readonly PersonID[];
  }): Promise<PersonTable[]> {
    const personRows = await params
      .knex<PersonTableRow>(Table.PERSONS)
      .whereIn("id", params.personIds);

    return personRows.map(formatPersonRow);
  },
};

type ApplyPersonSortResponse = {
  column: string;
  order: SortOrder;
};

function applyPersonSort(sort?: PersonSort[]): ApplyPersonSortResponse[] {
  if (!sort) {
    return [{ column: "firstName", order: SortOrder.Asc }];
  }

  return sort.map((element) => ({
    column: element.field,
    order: element.order,
  }));
}

export function addPersonFilters(
  queryBuilder: Knex.QueryBuilder,
  filterOperation?: PersonFilterOperation,
): Knex.QueryBuilder {
  return buildFilterQuery(queryBuilder, applyPersonFilters, filterOperation);
}

function applyPersonFilters(input: {
  queryBuilder: Knex.QueryBuilder;
  filterOperator: FilterOperator;
  filter: PersonFilter;
}): Knex.QueryBuilder {
  const { queryBuilder, filterOperator, filter } = input;
  if (filter.birthdayFilter) {
    applyDateFilters({
      queryBuilder,
      filterOperator,
      field: tableColumn(Table.PERSONS, "birthday"),
      dateFilter: filter.birthdayFilter,
    });
  }
  if (filter.nameFilter) {
    return applyStringFilters({
      queryBuilder,
      filterOperator,
      field: tableColumn(Table.PERSONS, "firstName"),
      stringFilter: filter.nameFilter,
    });
  }

  return queryBuilder;
}
