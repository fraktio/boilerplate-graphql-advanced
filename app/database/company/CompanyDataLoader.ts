import DataLoader from "dataloader";

import { companyQueries, CompanyID, CompanyTable } from "./companyQueries";

import {
  AbstractDataLoaderBase,
  DataLoaderParams,
} from "~/database/AbstractDataLoader";
import { Maybe } from "~/generation/generated";

export type CompanyLoader = DataLoader<CompanyID, Maybe<CompanyTable>>;

export class CompanyDataLoader extends AbstractDataLoaderBase<CompanyLoader> {
  protected createLoader(params: DataLoaderParams): CompanyLoader {
    const companyLoader: CompanyLoader = new DataLoader(async (ids) => {
      const companies = await companyQueries.getCompaniesByIds({
        knex: params.knex,
        companyIds: ids,
      });

      return ids.map(
        (id) => companies.find((company) => company.id === id) || null,
      );
    });

    return companyLoader;
  }
}
