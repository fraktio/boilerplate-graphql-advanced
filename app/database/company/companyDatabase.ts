import { DateTime } from "luxon";

import { Maybe } from "~/@types/global";
import { CompanyDataLoader } from "~/database/company/CompanyDataLoader";
import {
  companyQueries,
  CompanyID,
  CompanyTableRow,
} from "~/database/company/companyQueries";
import { DBSession } from "~/database/connection";
import { CompaniesOfPersonDataLoader } from "~/database/employee/CompaniesOfPersonDataLoader";
import { PersonFilterOperation } from "~/database/person/personFilters";
import { PersonID } from "~/database/person/personQueries";
import { UUID } from "~/generation/mappers";

export type CompanyTable = {
  id: CompanyID;
  UUID: UUID;
  name: string;
  timestamp: {
    createdAt: DateTime;
    updatedAt: DateTime | null;
  };
};

export const formatCompanyRow = (row: CompanyTableRow): CompanyTable => ({
  id: row.id,
  UUID: row.uuid,
  name: row.name,
  timestamp: {
    createdAt: DateTime.fromJSDate(row.createdAt),
    updatedAt: row.updatedAt ? DateTime.fromJSDate(row.updatedAt) : null,
  },
});

export const companyDB = {
  async get(params: {
    knex: DBSession;
    companyId: CompanyID;
    companyDL: CompanyDataLoader;
  }): Promise<Maybe<CompanyTable>> {
    const company = await companyQueries.get(params);

    if (company) {
      params.companyDL
        .getLoader({ knex: params.knex })
        .prime(company.id, company);
    }

    return company;
  },

  async tryGet(params: {
    knex: DBSession;
    companyId: CompanyID;
    companyDL: CompanyDataLoader;
  }): Promise<CompanyTable> {
    const company = await companyQueries.tryGet(params);

    params.companyDL
      .getLoader({ knex: params.knex })
      .prime(company.id, company);

    return company;
  },

  async getByUUID(params: {
    knex: DBSession;
    companyUUID: UUID;
    companyDL: CompanyDataLoader;
  }): Promise<Maybe<CompanyTable>> {
    const company = await companyQueries.getByUUID(params);

    if (company) {
      params.companyDL
        .getLoader({ knex: params.knex })
        .prime(company.id, company);
    }

    return company;
  },

  async getAll(params: {
    knex: DBSession;
    companyDL: CompanyDataLoader;
    filters?: PersonFilterOperation;
  }): Promise<CompanyTable[]> {
    const companies = await companyQueries.getAll(params);

    const dataloader = params.companyDL.getLoader({ knex: params.knex });
    companies.forEach((company) => {
      dataloader.prime(company.id, company);
    });

    return companies;
  },

  async create(params: {
    knex: DBSession;
    newCompany: { name: string };
    companyDL: CompanyDataLoader;
  }): Promise<CompanyTable> {
    const company = await companyQueries.create({
      knex: params.knex,
      company: params.newCompany,
    });

    params.companyDL
      .getLoader({ knex: params.knex })
      .prime(company.id, company);

    return company;
  },

  async updateByUUID(params: {
    knex: DBSession;
    companyUUID: UUID;
    company: { name: string };
    companyDL: CompanyDataLoader;
  }): Promise<Maybe<CompanyTable>> {
    const company = await companyQueries.updateByUUID(params);

    if (company) {
      params.companyDL
        .getLoader({ knex: params.knex })
        .prime(company.id, company);
    }

    return company;
  },

  async getCompaniesOfPerson(params: {
    knex: DBSession;
    personId: PersonID;
    companyDL: CompanyDataLoader;
    companiesOfPersonDL: CompaniesOfPersonDataLoader;
  }): Promise<CompanyTable[]> {
    const companyIds = await params.companiesOfPersonDL
      .getLoader({ knex: params.knex })
      .load(params.personId);

    const companies = await params.companyDL
      .getLoader({ knex: params.knex })
      .loadMany(companyIds);

    return companies as CompanyTable[];
  },
};
