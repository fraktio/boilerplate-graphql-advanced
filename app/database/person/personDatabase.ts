import { PersonsOfCompanyDataLoader } from "../employee/PersonsOfCompanyDataLoader";

import { PersonDataLoader } from "./PersonDataLoader";

import { CompanyID } from "~/database/company/companyQueries";
import { DBConnection } from "~/database/connection";
import {
  CreatePersonOptions,
  personQueries,
  PersonID,
  PersonTable,
  UpdatePersonOptions,
} from "~/database/person/personQueries";
import {
  Maybe,
  PersonFilterOperation,
  PersonSort,
} from "~/generation/generated";
import { UUID } from "~/generation/mappers";

export const personDB = {
  async get(params: {
    knex: DBConnection;
    personId: PersonID;
    personDL: PersonDataLoader;
  }): Promise<Maybe<PersonTable>> {
    return params.personDL
      .getLoader({ knex: params.knex })
      .load(params.personId);
  },

  async getByUUID(params: {
    knex: DBConnection;
    personUUID: UUID;
    personDL: PersonDataLoader;
  }): Promise<Maybe<PersonTable>> {
    const person = await personQueries.getByUUID(params);

    if (person) {
      params.personDL.getLoader({ knex: params.knex }).prime(person.id, person);
    }

    return person;
  },

  async getAll(params: {
    knex: DBConnection;
    personDL: PersonDataLoader;
    filters?: PersonFilterOperation;
    sort?: PersonSort[];
  }): Promise<PersonTable[]> {
    const persons = await personQueries.getAll(params);

    persons.forEach((person) => {
      const dataLoader = params.personDL.getLoader({ knex: params.knex });
      dataLoader.prime(person.id, person);
    });

    return persons;
  },

  async create(params: {
    knex: DBConnection;
    person: CreatePersonOptions;
    personDL: PersonDataLoader;
  }): Promise<PersonTable> {
    const person = await personQueries.create(params);

    params.personDL.getLoader({ knex: params.knex }).prime(person.id, person);

    return person;
  },

  async updateByUUID(params: {
    knex: DBConnection;
    personUUID: UUID;
    person: UpdatePersonOptions;
    personDL: PersonDataLoader;
  }): Promise<Maybe<PersonTable>> {
    const person = await personQueries.updateByUUID(params);

    if (person) {
      params.personDL.getLoader({ knex: params.knex }).prime(person.id, person);
    }

    return person;
  },

  async getPersonsOfCompany(params: {
    knex: DBConnection;
    companyId: CompanyID;
    personDL: PersonDataLoader;
    personsOfCompanyDL: PersonsOfCompanyDataLoader;
  }): Promise<PersonTable[]> {
    const personIds = await params.personsOfCompanyDL
      .getLoader({ knex: params.knex })
      .load(params.companyId);

    return (await params.personDL
      .getLoader({ knex: params.knex })
      .loadMany(personIds)) as PersonTable[];
  },
};
