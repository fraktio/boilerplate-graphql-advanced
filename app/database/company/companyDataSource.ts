import { DateTime } from "luxon";

import { CompanyDataLoader } from "./CompanyDataLoader";

import {
  companyDB,
  CompanyID,
  CompanyTableRow,
} from "~/database/company/companyDatabase";
import { DBConnection } from "~/database/connection";
import { PersonID } from "~/database/person/personDatabase";
import { UUID } from "~/generation/mappers";

export type CompanyTable = {
  id: CompanyID;
  UUID: UUID;
  name: string;
  timestamp: {
    createdAt: DateTime;
    updatedAt: DateTime | null;
  };
};

export const formatCompanyRow = (row: CompanyTableRow): CompanyTable => ({
  id: row.id,
  UUID: row.uuid,
  name: row.name,
  timestamp: {
    createdAt: DateTime.fromJSDate(row.createdAt),
    updatedAt: row.updatedAt ? DateTime.fromJSDate(row.updatedAt) : null,
  },
});

export const companyDS = {
  async get(params: {
    knex: DBConnection;
    companyId: CompanyID;
    companyDL: CompanyDataLoader;
  }) {
    const company = await companyDB.get(params);

    if (company) {
      params.companyDL
        .getLoader({ knex: params.knex })
        .prime(company.id, company);
    }

    return company;
  },

  async tryGet(params: {
    knex: DBConnection;
    companyId: CompanyID;
    companyDL: CompanyDataLoader;
  }) {
    const company = await companyDB.tryGet(params);

    params.companyDL
      .getLoader({ knex: params.knex })
      .prime(company.id, company);

    return company;
  },

  async getByUUID(params: {
    knex: DBConnection;
    companyUUID: UUID;
    companyDL: CompanyDataLoader;
  }) {
    const company = await companyDB.getByUUID(params);

    if (company) {
      params.companyDL
        .getLoader({ knex: params.knex })
        .prime(company.id, company);
    }

    return company;
  },

  async getAll(params: { knex: DBConnection; companyDL: CompanyDataLoader }) {
    const companies = await companyDB.getAll(params);

    const dataloader = params.companyDL.getLoader({ knex: params.knex });
    companies.forEach((company) => {
      dataloader.prime(company.id, company);
    });

    return companies;
  },

  async create(params: {
    knex: DBConnection;
    newCompany: { name: string };
    companyDL: CompanyDataLoader;
  }) {
    const company = await companyDB.create({
      knex: params.knex,
      company: params.newCompany,
    });

    params.companyDL
      .getLoader({ knex: params.knex })
      .prime(company.id, company);

    return company;
  },

  async updateByUUID(params: {
    knex: DBConnection;
    companyUUID: UUID;
    company: { name: string };
    companyDL: CompanyDataLoader;
  }) {
    const company = await companyDB.updateByUUID(params);

    if (company) {
      params.companyDL
        .getLoader({ knex: params.knex })
        .prime(company.id, company);
    }

    return company;
  },

  async getCompaniesOfPerson(params: {
    knex: DBConnection;
    personId: PersonID;
  }) {
    return await companyDB.getCompaniesOfPerson(params);
  },
};
