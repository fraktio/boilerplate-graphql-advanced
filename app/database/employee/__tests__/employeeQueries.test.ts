import { CompanyID } from "~/database/company/companyQueries";
import {
  EmployeeID,
  EmployeeTable,
  EmployeeTableRow,
  formatEmployeeRow,
} from "~/database/employee/employeeQueries";
import { PersonID } from "~/database/person/personQueries";

const datatimeString =
  "Fri Mar 12 2021 13:16:56 GMT+0200 (Eastern European Standard Time)";

export const dbEmployeeMockTableRow: EmployeeTableRow = {
  id: 2 as unknown as EmployeeID,
  companyId: 3 as unknown as CompanyID,
  personId: 5 as unknown as PersonID,
  createdAt: new Date(datatimeString),
};

export const createFakeEmployeeTable = (): EmployeeTable =>
  formatEmployeeRow(dbEmployeeMockTableRow);

describe("database tests / employee", () => {
  it("formatEmployeeRow", async () => {
    const result = formatEmployeeRow(dbEmployeeMockTableRow);

    expect(result).toMatchSnapshot();
  });
});
