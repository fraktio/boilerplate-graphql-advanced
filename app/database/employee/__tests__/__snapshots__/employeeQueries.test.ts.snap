// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`database tests / employee formatEmployeeRow 1`] = `
Object {
  "companyId": 3,
  "id": 2,
  "personId": 5,
  "timestamp": Object {
    "createdAt": "2021-03-12T11:16:56.000+00:00",
    "updatedAt": null,
  },
}
`;
